### ISSUE TITLE
Permission denied - AudioDispatcherFactory.fromPipe

### ISSUE THREAD

**Issue by blaster92** [JorenSix/TarsosDSP#176]
Hello everyone, I'm facing a problem about analysis of an audio track by using this library.
I'm using AudioDispatcherFactory.fromPipe function to fetch input data and, when running the code over a Xiaomi Mi 9, I receive this error message:

"/data/user/0/ismb.musikoach/cache/ffmpeg" -ss 0.0   -i "/storage/emulated/0/music/aaa.mp3" -vn -ar 44100 -ac 1 -sample_fmt s16 -f s16le pipe:1
I/PipeDecoder: /system/bin/sh: /data/user/0/isda.musss/cache/ffmpeg: Permission denied

Then, the handlers that evaluate music params return null.

I suppose it is a phone related problem, about the permission granted by the app to the phone itself. Anyway, I ensured that writing/reading permissions are granted to the app.

Thank you

Sergio

**Comment by tim162** [comment]
@blaster92 I'm currently having the same problem. Did you figure out how to solve the issue?

**Comment by pupil72** [comment]
2 years later and still the same problem: I can't process sound anymore with .fromPipe method after Android 11 came up. Anyone with the a solution?

**Comment by H-Fontaine** [comment]
Don't know if you still need it but I managed to make it work : https://github.com/H-Fontaine/Android-TarsosDSP

**Comment by pupil72** [comment]
Thanks a lot, I'll try to make it work! Although not essential for my app any longer, it would add a feature a user asked for.

If I understand correctly from your code, it's just question of adding the Arthenica library as a dependency and... that's it, .frompipe will work, isn't it? Am I missing something?

**Comment by H-Fontaine** [comment]
Exactly. The way FFmpeg was used wasn't working anymore because of security patchs so I took the Arthenica library to execute the command instead.

I put the library binaries in release.

**Comment by pupil72** [comment]
Hi Hugo, sorry I have 3 more questions if you have time to answer them:

1) I don't understand: "I put the library binaries in release." I look for your binaries but I can't find them. Would you mind to be more explicit about where to find and put them?
2) do you know what file type .frompipe accepts? .wav don't seem to work, but .mp3 do. Do you know about more extensions?
3) .fromPipe plays sound only for a few seconds, then stop. Any clue?

**Comment by H-Fontaine** [comment]
No problems,

1. I was talking about the .aar files that you can find there : https://github.com/H-Fontaine/Android-TarsosDSP/releases/tag/1.0

2. I have tested .MP3, .ACC, .AIFF, .FLAC, .OGG and .WAV and all works, at least I can play them. If I can play them I think everything else should work.

3. The javadoc of .frompipe, so to read the entire file you should give begin = 0.0 and duration = -1.0
	/**
	 * Create a stream from a piped sub process and use that to create a new
	 * {@link AudioDispatcher} The sub-process writes a WAV-header and
	 * PCM-samples to standard out. The header is ignored and the PCM samples
	 * are are captured and interpreted. Examples of executables that can
	 * convert audio in any format and write to stdout are ffmpeg and avconv.
	 *
	 * @param context
	 * 			  The context from witch comes the Uri path
	 * @param begin
	 * 			  Start time from which we read the file in seconds
	 * @param duration
	 * 			  Number of seconds read from the start time (begin) if negative read the entire file
	 * @param selectedFileUri
	 *            The file or stream to capture.
	 * @param targetSampleRate
	 *            The target sample rate.
	 * @param audioBufferSize
	 *            The number of samples used in the buffer.
	 * @param bufferOverlap
	 * 			  The number of samples to overlap the current and previous buffer.
	 * @return A new audioprocessor.
	 */
	public static AudioDispatcher fromPipe(Context context, Uri selectedFileUri, double begin, double duration, final int targetSampleRate, final int audioBufferSize, final int bufferOverlap)

I hope that will help you !

**Comment by pupil72** [comment]
Thanks again, but... after a couple of hours, oh my God! not even from microphone is working now!!
Let's see if I can show you what I did:

![Screenshot_1](https://user-images.githubusercontent.com/37589858/215287651-cebea896-a1c1-4940-b2ff-6d1dba4e1875.png)

1) I created a libs directory
2) copied there .aar files
3) added dependencies in Gradle

In the structure you can see that I put the whole Tarsos inside the app (not a good idea, I guess), I should add it as a module to the project as you did. I will, but anyway... I don't think this is why I get the following message when trying to record from microphone or process .fromPipe.

java.lang.NoSuchMethodError: No static method fromDefaultMicrophone(III)Lbe/tarsos/dsp/AudioDispatcher; in class Lbe/tarsos/dsp/io/android/AudioDispatcherFactory; or its super classes (declaration of 'be.tarsos.dsp.io.android.AudioDispatcherFactory' appears in /data/data/marcoalunno.optical_theremin/code_cache/.overlay/base.apk/classes.dex)

Your project works, though. I'm still trying to figure out why, but I don't see a libs folder with .aar inside

**Comment by H-Fontaine** [comment]
Ahh not like that ! 😆
The .aar files are meant to simplify your life ! It is made for you not to add all the files in of library everything is already included in the .aar files !

Normaly there already is a libs folder located at "You project"/app/libs (on your disk, not in android studio the ide show you something else (depends on the view you use)). You put the .aar files there and add the 3 lines your dependencies as you did and sync gradle.
After that you are good to go ! You just have to make import at the beginning of your java file as you normaly do !

**Comment by pupil72** [comment]
No, I know that!
1) the libs directory should be there, but it has never happened. I have to create it all the time I need it.
2) the image shows you the structure of the project no the app view

You can see here that the .aar are in the path "Theremin_146/app/libs"
![image](https://user-images.githubusercontent.com/37589858/215291360-57632a00-9bbb-452a-a866-b15ebb68bf93.png)

Well... Hugo, I don't want to waste your time. I'll keep trying to undestand why your example works and my doesn't!

**Comment by H-Fontaine** [comment]
I hope you manage to make it work ! Good luck !

**Comment by pupil72** [comment]
Hi Hugo. Just to let you know that I made it work!! I basically reconstructed my app around yours. Not only my dumbass ignorance in importing a simple module (which is what you try to let me understand several times), I also had problems with: 1) custom methods I had made in TarsosDSP previously which didn't work any longer with your implementation of Tarsos; 2) a gradle upgrade that destroyed my app; 3) dependencies that didn't work anymore with the new gradle, etc....

But now it works again! and I can use fromPipe!! ...although it sounds quite distorted... I'll try to figure out why...
